src/orion/algo/robo/ablr/ablr_model.py

```diff
@@ -0,0 +1,589 @@
+        state_dict["optimizer"] = self.optimizer.state_dict()
+        return state_dict
+
+    def load_state_dict(self, state_dict: dict, strict: bool = True) -> tuple:
+        optim_state: dict = state_dict.pop("optimizer")
+        self.optimizer.load_state_dict(optim_state)
+        rng: Tensor = state_dict.pop("rng")
+        torch.random.set_rng_state(rng)
+        return super().load_state_dict(state_dict, strict=strict)  # type: ignore
+
+    def set_state(self, state_dict: dict) -> None:
+        """Orion hook used to restore the state of the algorithm."""
+        self.load_state_dict(state_dict)
+
+    def seed(self, seed: int) -> None:
+        """Seed the state of the random number generator.
+
+        :param seed: Integer seed for the random number generator.
+
+        .. note:: This methods does nothing if the algorithm is deterministic.
+        """
+        # No need to do anything here really, since the ROBO_ABLR class already seeds all the
+        # pytorch RNG.
+
+    def predict(self, X_test: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        """Return the predictive mean and variance for the given points."""
+        y_pred_distribution = self.predict_dist(X_test)
+        return (
+            y_pred_distribution.mean.cpu().numpy(),
+            y_pred_distribution.variance.cpu().numpy(),
+        )
+
+    def predict_dist(self, x_test: Tensor | np.ndarray) -> Normal:
+        """Return the predictive distribution for the given points."""
+        with torch.no_grad():
+            if self._predict_dist is None:
+                # re-create the predictive distribution.
+                # NOTE: This gets overwritten every time a new point is observed, so we know for
+                # sure that this is always the most up-to-date predictions.
+                _, self._predict_dist = self(self.X, self.y)
+            y_pred_dist = self._predict_dist(x_test)
+        return y_pred_dist
+
+    def forward(
+        self, x_train: Tensor, y_train: Tensor
+    ) -> tuple[Tensor, Callable[[Tensor], Normal]]:
+        """Return the training loss and a function that will give the
+        predictive distribution over y for a given (un-normalized) input x_test.
+
+        NOTE: Assumes that `x_train`, `y_train` and an eventual `x_test` are
+        NOT already rescaled/normalized!
+        """
+        x_train = torch.as_tensor(x_train)
+        y_train = torch.as_tensor(y_train)
+        # TODO: Do we normalize the portion that has the contextual information?
+        if all(self.x_mean == 0):
+            self.x_mean = x_train.mean(0)
+        if all(self.x_var == 1):
+            x_var = x_train.var(0)
+            # minimum variance, because the context vectors are actually always
+            # the same, so the variance along that dimension is 0, which makes
+            # NaN values in the normalized inputs.
+            self.x_var = torch.maximum(x_var, torch.zeros_like(x_var) + 1e-5)
```

Comment:
Is forward call on all data at once? Why is the variance compute on forward once only?

Reply:
Ooh nice catch... I don't recall! I'll check.

Reply:
Ok, the mean and variance of the entire training dataset are used, and they are being set externally in the ABLR model.